packages/js/product-editor/src/components/details-name-block/edit.tsx

```diff
@@ -2,36 +2,175 @@
+} from '@woocommerce/data';
+import {
+	Button,
+	BaseControl,
+	// @ts-expect-error `__experimentalInputControl` does exist.
+	__experimentalInputControl as InputControl,
+} from '@wordpress/components';
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment
 // @ts-ignore No types for this exist yet.
 // eslint-disable-next-line @woocommerce/dependency-group
-import { useEntityProp } from '@wordpress/core-data';
+import { useEntityProp, useEntityId } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { AUTO_DRAFT_NAME } from '../../utils';
+import { EditProductLinkModal } from '../edit-product-link-modal';
+import { useValidation } from '../../hooks/use-validation';
+import { useProductHelper } from '../../hooks/use-product-helper';
 
 export function Edit() {
 	const blockProps = useBlockProps();
-	const [ name, setName ] = useEntityProp( 'postType', 'product', 'name' );
+
+	const { updateProductWithStatus } = useProductHelper();
+
+	const [ showProductLinkEditModal, setShowProductLinkEditModal ] =
+		useState( false );
+
+	const { editEntityRecord } = useDispatch( 'core' );
+	const productId = useEntityId( 'postType', 'product' );
+	const product: Product = useSelect( ( select ) =>
+		select( 'core' ).getEditedEntityRecord(
+			'postType',
+			'product',
+			productId
+		)
+	);
+
+	const [ name, setName ] = useEntityProp< string >(
+		'postType',
+		'product',
+		'name'
+	);
+
+	const { permalinkPrefix, permalinkSuffix } = useSelect(
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		( select: WCDataSelector ) => {
+			const { getPermalinkParts } = select( PRODUCTS_STORE_NAME );
+			if ( productId ) {
+				const parts = getPermalinkParts( productId );
+				return {
+					permalinkPrefix: parts?.prefix,
+					permalinkSuffix: parts?.suffix,
+				};
+			}
+			return {};
+		}
+	);
+
+	const nameIsValid = useValidation(
+		'product/name',
+		() => Boolean( name ) && name !== AUTO_DRAFT_NAME
+	);
 
 	return (
-		<div { ...blockProps }>
-			<TextControl
-				label={ createInterpolateElement(
-					__( 'Name <required />', 'woocommerce' ),
-					{
-						required: (
-							<span className="woocommerce-product-form__optional-input">
-								{ __( '(required)', 'woocommerce' ) }
-							</span>
-						),
-					}
+		<>
+			<div { ...blockProps }>
+				<BaseControl
+					id={ 'product_name' }
+					label={ createInterpolateElement(
+						__( 'Name <required />', 'woocommerce' ),
+						{
+							required: (
+								<span className="woocommerce-product-form__required-input">
+									{ __( '*', 'woocommerce' ) }
+								</span>
+							),
+						}
+					) }
+				>
+					<InputControl
+						name={ 'woocommerce-product-name' }
+						placeholder={ __(
+							'e.g. 12 oz Coffee Mug',
+							'woocommerce'
+						) }
+						onChange={ setName }
+						value={ name || '' }
+					/>
+				</BaseControl>
+				{ productId &&
+					nameIsValid &&
+					product.status === 'publish' &&
```

Comment:
The product link should also show up when the product is in draft, it is only not visible when in `auto-draft`.